src/si/fmt.rs

```diff
@@ -0,0 +1,133 @@
+use lib::fmt;
+
+use si::{Dimension, Quantity, Unit, Units};
+use num::Num;
+use {Conversion, ConversionFactor};
+
+/// Specifies the display style to use.
+#[derive(Clone, Copy, Debug)]
+pub enum DisplayStyle {
+    /// Display the value and a unit abbreviation, e.g. "1.0 m", "327 s".
+    Abbreviation,
+    /// Display the value and full unit name (pluralized as appropriate), e.g. "1 kilogram", "756 feet".
+    Description,
+}
+
+/// Provides an interface to specify a display style and unit.
+///
+/// # Usage
+/// ```
+/// # use uom::si::Unit;
+/// # use uom::si::{f32::*, length::{centimeter, meter}, fmt::{Arguments, DisplayStyle::*}};
+/// let length = Length::new::<meter>(1.0);
+/// let a = Arguments::new(centimeter, Description);
+/// assert_eq!("100 centimeters", format!("{}", a.with(length)));
+/// ```
+#[allow(missing_debug_implementations)] // Prevent accidental direct use.
+#[derive(Clone)]
+pub struct Arguments<T: Unit> {
+    subunit: T,
+    style: DisplayStyle,
+}
+
+impl<T: Unit> Arguments<T> {
+
+    /// Creates a new `Argument` for displaying a quantity in the specified units and style.
+    pub fn new(subunit: T, style: DisplayStyle) -> Self {
+        Self { subunit, style }
+    }
+
+    /// Specifies a quantity to display.
+    pub fn with<D, U, V>(&self, quantity: Quantity<D, U, V>) -> QuantityArguments<T, D, U, V>
+    where
+        D: Dimension + ?Sized,
+        U: Units<V> + ?Sized,
+        V: Num + Conversion<V>,
+    {
+        QuantityArguments {
+            arguments: self.clone(), // Allow use of this format style with multiple quantities
+            quantity,
+        }
+    }
+
+}
+
+#[doc(hidden)]
+pub struct QuantityArguments<T: Unit, D: Dimension + ?Sized, U: Units<V> + ?Sized, V: Num + Conversion<V>> {
+    arguments: Arguments<T>,
+    quantity: Quantity<D, U, V>,
+}
+
+trait Convert<V> { // Use this get<N> instead of the (private) get<N> implemented per-unit.
+    fn get<N: Unit + Conversion<V>>(&self, _unit: N) -> V;
+}
+
+impl<D, U, V> Convert<V> for Quantity<D, U, V>
+where
+    D: Dimension + ?Sized,
+    U: Units<V> + ?Sized,
+    V: Num + Conversion<V> + Clone,
+{
+    fn get<T: Unit + Conversion<V>>(&self, _unit: T) -> V {
+        self.value.clone() / T::conversion().value()
+    }
+}
+
+macro_rules! display_arguments {
+    ($name:ident, $pat:expr) => {
+        impl<T, D, U, V> fmt::$name for QuantityArguments<T, D, U, V>
+        where
+            T: Unit + Conversion<V>,
+            D: Dimension + ?Sized,
+            U: Units<V> + ?Sized,
+            V: Num + Conversion<V> + fmt::$name + Clone,
+        {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                let value = self.quantity.get::<T>(self.arguments.subunit);
+                write!(f, $pat, value)?;
```

Comment:
Instead of passing `$pat` in via macro parameter call the value's `fmt::$name` implementation: `value.fmt(f)` (or, if explicit call syntax is needed: `<V as ::lib::fmt::$name>::fmt(v, f)`). See the `Debug` impl for `Quantity` for an example.